### ISSUE TITLE
Overriding deepcopy in Python API

### ISSUE THREAD

**Issue by paulromano** [mit-crpg/openmc#659]
Many of our classes override `__deepcopy__` to provide a custom implementation. However, as far as I can tell, the custom implementation doesn't do anything special that a default deepcopy wouldn't do in many of these cases. One reason I can think of for wanting to override `__deepcopy__` is to automatically generate a new `id`, i.e. the copy of a `Cell` would not have the same `id` as the original. @wbinventor Do you think deepcopies of objects that have an `id` attribute should get the same `id` or a new one?

It does look like the MGXS-related custom deepcopies are indeed custom (some attributes are shallow copied). Not sure if they have to be though. Maybe you can shed more light on that @wbinventor.

**Comment by wbinventor** [comment]
I'll try to explain the rationale behind the `__deepcopy__` methods, but it has been quite a while since I implemented this code so I may not do the explanation justice. I've tried to split my comments along the roughly two axes you mentioned @paulromano:

**Shallow vs. Deep Copies of Class Attributes**
It is true that most of the `__deepcopy__` methods simply perform deep copies of each class attribute as would be done with the default `__deepcopy__`. The only difference are the few class attributes which are only shallow copied. Based on my most recent cursory overview, it appears that this only a factor for the `openmc.mgxs` module (as you point out @paulromano), so I suppose that many of the `__deepcopy__` method elsewhere could simply be removed and replaced by the default. For those implementations which make use of shallow copies of object references (*i.e.*, the `MGXS` shallow copies its `Cell` or `Material` domain) in order to ensure that objects remain "linked" together as specified by the user. For example, if you build a geometry and specifies a domain for an `MGXS`, and then deep copies that `MGXS`, you would (likely) expect the `MGXS` copy to still point to the same domain. This would allow you to make any changes you like to that domain, and both the original and the copied `MGXS` will point to the same changed domain. This might sound like a far-fetched use case, but is pretty common in a lot of the simulation pipelines I use.

**Generating Object IDs**
The reason the IDs are not (re-)generated for deep copied is at its heart due to the legacy of the original implementations of everything in OpenCG. In particular, the "region differentiation" algorithm (which is the primary value of OpenCG to everything I do) in OpenCG requires that the IDs remain static with a deep copy. I did implement a `clone` method for each object in OpenCG which *does* generate a new ID. In retrospect the `__deepcopy__`/`clone` combo may be a bit clunky. Perhaps we could introduce an optional boolean `generate_id` parameter to the `__deepcopy__` methods in the OpenMC Python API in order to permit the type of behavior you desire @paulromano, while still supporting more novel CSG algorithms like region differentiation (upon which I will depend for at least a few more months yet). Of course I'm the only person who is using region differentiation, so feel free to ignore my plea if you wish!

**Comment by paulromano** [comment]
**Regarding MGXS** -- why not just do a shallow copy if you want to ensure the objects remain linked?

**Regarding object IDs** -- unfortunately an extra argument on `__deepcopy__` won't be usable because the method signature for `copy.deepcopy` doesn't pass anything to the underlying `__deepcopy__` implementation other than the object and the memo dictionary. So, I think you'd still have to roll your own method (like `clone`) in order to have that level of customization. I guess I'd prefer to see OpenMC's `__deepcopy__` methods generate a new ID by default and have `clone` be responsible for deepcopies with identical IDs.

**Comment by paulromano** [comment]
I've created a [deepcopy](https://github.com/paulromano/openmc/tree/deepcopy) branch where I've gotten rid of most of the existing `__deepcopy__` implementations and rewritten ones for classes with IDs. For the MGXS issue, I create shallow copies of libraries, which seems to work? Let me know what you think @wbinventor.

**Comment by wbinventor** [comment]
I don't think that the shallow copies of libraries will work as desired for group condensation or subdomain-averaging. Won't that simply condense/average the original `Library` rather than apply the operations to a new one?

Also, I'm curious - do you have a use case in mind where deep copying with new IDs would be advantageous? I really don't see how this is a *good* thing, and find it kind of counter intuitive to what one (or at least I) would expect from a deep copy. 

There is at least one example where I *think* all of my code for my thesis will fail with your deep copy branch as a result of the new IDs. I deep copy MGXS `Library` objects all over the place and do different subdomain-averaging and energy condensation operations on them. In each case, I change the OpenMC `Geometry` pointed to by the is "differentiated" to look like the subdomain-averaging operation applied to the distribcell MGXS (i.e., make a geometry that looks like "clustered" distribcell data). If the IDs are different in each deep copied `Geometry` all of this will fall apart. 

The only other person I know who has explicitly mentioned the convenience of deep copying objects and retaining their IDs is @nhorelik, though he has likely forgotten why he wanted this now that he's off wandering greener pastures. But in any case, I do recall a conversation with him when he mentioned that this was helpful in generating lots of different cases to run for this thesis.

Perhaps our use case(s) can be refactored to work with the new proposed deep copies. But I'm confident that it is very convenient to have deep copies retain the original IDs if you are generating lots and lots of inputs derived off an original one. This plea isn't in defense of the current `__deepcopy__` implementations - you've raised a good point that we can do away with many of the implementations altogether in favor of the default. At the end of the day, I'm going to have to fix my OpenMC repo and go off the grid anyway.

**Comment by paulromano** [comment]
My original use case was related to #667 -- to place TRISO particles in a lattice, if they overlap a lattice boundary they need to be duplicated in each lattice element which might contain them. In this case, I needed to duplicate each cell defining a layer of the TRISO, and since the surfaces would be translated based on the lattice element origin, the surfaces would be duplicated as well. Even so, I didn't really want a full deepcopy because the materials should stay the same. In the end, I decided to just hard reset the IDs of the surfaces/cells via `id = None`.

For now, #667 at least improves the situation by removing `__deepcopy__` implementations that essentially mimic the default behavior. I'm not sure what I find to be intuitive or counter-intuitive: on one hand the deepcopy should produce an object with identical (in value) instance variables, but on the other hand, the "unique" ID is no longer necessarily unique.

**Comment by wbinventor** [comment]
That's an interesting use case - I can see why you would want to adjust the IDs in that situation. There certainly is no "right" answer here. I recognize that the infrastructure I've developed to support my use case makes the implicit assumption that IDs are not re-generated, but that the code could likely be modified to work with newly generated IDs. That said, my personal intuition for a deep copy would lead me to expect it to "produce an object with identical (in value) instance variables" but that's just my perspective.

While we're on the subject of "unique" IDs, this conversation is reminiscent of a lot of conversations I had with Nick and others about "unique" ID generation in OpenCG back in the day. I certainly don't think we came up with the "best" solution, but our approach is a bit of a hybrid. Each object has two IDs - one which is unique to the global namespace (these do change in a deep copy), along with integer IDs which are assigned by the user or auto-generated by a `Geometry` (these do not change in a deep copy). I believe that this approach would support both of our use cases.
